oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedMongoServerSelector.java

```diff
@@ -0,0 +1,179 @@
+                .collect(Collectors.joining("\n", "\n", "")));
+        var secondaries = serverDescriptions.stream()
+                .filter(ServerDescription::isSecondary)
+                .collect(Collectors.toSet());
+        if (secondaries.isEmpty()) {
+            LOG.info("No secondaries found, not selecting the primary. Returning empty list.");
+            return List.of();
+        }
+        // Secondaries available
+        serverAddressHashMap.remove(threadId);
+        // No server assigned to this thread yet. Find an unused server
+        for (var replica : secondaries) {
+            var address = replica.getAddress();
+            if (!serverAddressHashMap.containsValue(address)) {
+                // This server is not assigned to any thread yet. Assign it to this thread
+                serverAddressHashMap.put(threadId, address);
+                LOG.info("Selected server: {}", address);
+                return List.of(replica);
+            }
+        }
+        // All secondaries are already assigned to some thread
+        LOG.debug("All available Mongo secondaries are assigned. Returning empty list.");
+        return List.of();
+    }
+
+    @Override
+    public void clusterOpening(ClusterOpeningEvent event) {
+        // Ignore
+    }
+
+    @Override
+    public synchronized void clusterClosed(ClusterClosedEvent event) {
+        // Ignore
+    }
+
+    @Override
+    public synchronized void clusterDescriptionChanged(ClusterDescriptionChangedEvent event) {
+        // Secondaries can be promoted to primary at any time. If that happens, any downloader thread connected to this
+        // replica should disconnect. We use the cluster description event to detect if any of the threads is connected
+        // to the replica that is now the primary. The downloader threads are then responsible for periodically calling
+        // #isConnectedToPrimary to check if they are connected to the primary and in that case close the connection
+        // and open a new connection, which will trigger the Mongo driver to call again the selection logic in this class
+        this.lastSeenClusterDescription = event.getNewDescription();
+        if (lastSeenClusterDescription.getType() != ClusterType.REPLICA_SET) {
+            return;
+        }
+
+        // Check if any thread is connected to primary
+        connectedToPrimaryThreads.clear();
+        lastSeenClusterDescription.getServerDescriptions().stream()
+                .filter(ServerDescription::isPrimary)
+                .map(ServerDescription::getAddress)
+                .forEach(primaryAddress -> {
+                    for (var entry : serverAddressHashMap.entrySet()) {
+                        if (entry.getValue().equals(primaryAddress)) {
+                            connectedToPrimaryThreads.add(entry.getKey());
+                        }
+                    }
+                });
```

Comment:
to be consistent in terms of style, you can write this logic in a fully functional way. This would require `connectedToPrimaryThreads` to be set as non final though. It's a matter of taste.

```
private Set<Long> connectedToPrimaryThreads = new HashSet<>();
```

```suggestion
        connectedToPrimaryThreads = lastSeenClusterDescription.getServerDescriptions().stream()
    .filter(ServerDescription::isPrimary)
    .map(ServerDescription::getAddress)
    .flatMap(primaryAddress -> serverAddressHashMap.entrySet().stream()
        .filter(entry -> primaryAddress.equals(entry.getValue()))
        .map(Map.Entry::getKey))
    .collect(Collectors.toSet());
```

Reply:
I initially wrote the code all as streams, but I did not like the stream-within-a-stream nesting and the use of flatMap. I feel that the forEach loop is a bit more clear.